src/Modal/Modal.tsx

```diff
@@ -0,0 +1,212 @@
+import * as React from 'react';
+import styled, { keyframes } from 'styled-components';
+import clsx from 'clsx';
+import Transition, { TransitionStatus } from 'react-transition-group/Transition';
+import ModalOverlay from './ModalOverlay';
+import { Box } from '../Box';
+import { Portal } from '../Portal';
+import { ANIMATION_DURATION } from './constants';
+
+export type ModalSizes = 'sm' | 'md' | 'lg';
+
+export interface ModalProps {
+  /** Additional CSS classes to give to the modal. */
+  className?: string;
+  /** Additional CSS properties to give to the modal. */
+  style?: React.CSSProperties;
+  /** Whether the modal is open or not. */
+  isOpen: boolean;
+  /** Set to `true` if you want to hide the modal backdrop. */
+  noBackdrop?: boolean;
+  /** Hides the default close button. Useful if you want to add custom close behaviour. */
+  hideCloseButton?: boolean;
+  /** Set to `true` to disable closing the modal by clicking the overlay. */
+  disableOverlayClick?: boolean;
+  /**
+   * Enables focus trap mode.
+   * @TODO implement focus trapping: https://www.npmjs.com/package/focus-trap
+   */
+  enableFocusTrap?: boolean;
+  /** Used to reference the ID of the title element in the modal */
+  labelledById?: string;
+  /** Set max size of the modal */
+  size?: ModalSizes;
+  /** Callback method run when the Close button is clicked. */
+  onClose?: () => void;
+}
+
+interface ModalState {
+  isOpen: boolean;
+}
+
+const ModalIn = keyframes`
+  0% {
+    opacity: 0;
+    transform: translate(0, -25%);
+  }
+  100% {
+    opacity: 1;
+    transform: translateY(0);
+  }
+`;
+
+const ModalOut = keyframes`
+  0% {
+    opacity: 1;
+    transform: translateY(0);
+  }
+  100% {
+    opacity: 0;
+    transform: translate(0, -25%);
+  }
+`;
+
+const ModalContent = styled(Box)`
+  opacity: 0;
+  transform: translate(0, -25%);
+  overflow: hidden;
+  &[data-state='entering'],
+  &[data-state='entered'] {
+    animation-fill-mode: forwards;
+    animation-name: ${ModalIn};
+    animation-duration: ${ANIMATION_DURATION}ms;
+  }
+  &[data-state='exiting'] {
+    animation-fill-mode: forwards;
+    animation-name: ${ModalOut};
+    animation-duration: ${ANIMATION_DURATION}ms;
+  }
+  &.entered {
+    opacity: 1;
+    transform: translate(0, 0);
+  }
+`;
```

Comment:
Should we bring all these guys into a `styled.ts` file beside the component? I've done it for the Input/Field but still open to changing